File: 1071-greatest-common-divisor-of-strings/1071-greatest-common-divisor-of-strings.java
class Solution {
    public String gcdOfStrings(String str1, String str2) {
        int l1 = str1.length();
        int l2 = str2.length();

        for(int i= Math.min(l1, l2); i>0; i--) {
            if( valid(i, str1, str2, l1, l2) ){
                return str1.substring(0,i);
            }

            
        }

        return "";
    }

    public Boolean valid(int i, String str1, String str2, int l1, int l2) {
        if(l1 % i > 0 || l2 % i > 0) {
            return false;
        }


        String base = str1.substring(0,i);

        return str1.replace(base, "").isEmpty() && str2.replace(base, "").isEmpty();
    }
}

File: 1071-greatest-common-divisor-of-strings/1071-greatest-common-divisor-of-strings.py
class Solution:
    def gcdOfStrings(self, str1: str, str2: str) -> str:
        
        l = min(len(str1), len(str2))
        
        while l>0:
            
            if self.helper(l, str1, str2):
                return str1[:l]
            
            l-=1
        
        return ""
    
    
    def helper(self, l, str1, str2):
        
        if len(str1)%l>0 or len(str2)%l>0:
            return False
        
        base = str1[:l]
        
        n1, n2 = len(str1)//l, len(str2)//l
        
        return n1*base == str1 and n2*base==str2
            
            